quint/src/types/constraintGenerator.ts

```diff
@@ -351,48 +372,88 @@ export class ConstraintGeneratorVisitor implements IRVisitor {
     return this.solvingFunction(this.table, constraint)
       .mapLeft(errors => errors.forEach((err, id) => this.errors.set(id, err)))
       .map(subs => {
-        // For every free name we are binding in the substitutions, the names occuring in the value of the substitution
+        // For every free name we are binding in the substitutions, the names occurring in the value of the substitution
         // have to become free as well.
         this.addBindingsToFreeNames(subs)
 
         // Apply substitution to environment
         // FIXME: We have to figure out the scope of the constraints/substitutions
         // https://github.com/informalsystems/quint/issues/690
-        this.types = new Map<bigint, TypeScheme>(
-          [...this.types.entries()].map(([id, te]) => {
-            const newType = applySubstitution(this.table, subs, te.type)
-            const scheme: TypeScheme = this.quantify(newType)
-            return [id, scheme]
-          })
-        )
+        this.types.forEach((oldScheme, id) => {
+          const newType = applySubstitution(this.table, subs, oldScheme.type)
+          const newScheme: TypeScheme = this.quantify(newType)
+          this.addToResults(id, right(newScheme))
+        })
 
         return subs
       })
   }
 
-  private typeForName(name: string, nameId: bigint, arity: number): Either<ErrorTree, QuintType> {
+  // If there is a builtin called `name`, compute a new instance of its type scheme,
+  // otherwise, compute a new instance of the scheme previously computed for `id`
+  private typeForName(name: string, id: bigint, arity: number): Either<ErrorTree, QuintType> {
     // Assumes a valid number of arguments
 
     if (this.builtinSignatures.has(name)) {
       const signatureFunction = this.builtinSignatures.get(name)!
       const signature = signatureFunction(arity)
       return right(this.newInstance(signature))
     } else {
-      const def = this.table.get(nameId)
+      return this.typeForId(id).mapLeft(
+        (child: ErrorTree): ErrorTree => ({
+          location: `looking up type for ${name}`,
+          children: [child],
+        })
+      )
```

Comment:
We have a utility for this

```suggestion
      return this.typeForId(id).mapLeft(e => buildErrrorTree(`looking up type for ${name}`, e)
```